qgym/benchmarks/metrics/initial_mapping_metrics.py

```diff
@@ -0,0 +1,146 @@
+"""Metrics to assess the performance of InitialMappers.
+
+Functions named as ``*_score`` return a scalar value to maximize: the higher
+the better.
+
+Function named as ``*_error`` or ``*_loss`` return a scalar value to minimize:
+the lower the better.
+"""
+
+from __future__ import annotations
+
+from collections import deque
+from collections.abc import Iterable
+from copy import deepcopy
+from typing import Protocol, cast, runtime_checkable
+
+import networkx as nx
+import numpy as np
+from numpy.typing import ArrayLike, NDArray
+from qiskit import QuantumCircuit
+from qiskit.dagcircuit import DAGCircuit
+
+from qgym.generators.graph import BasicGraphGenerator, GraphGenerator
+from qgym.utils.input_validation import check_string
+
+
+@runtime_checkable
+class InitialMappingMetric(Protocol):
+    def compute(self, interaction_graph: nx.Graph, mapping: ArrayLike) -> float: ...
+
+
+@runtime_checkable
+class Mapper(Protocol):
+    def compute_mapping(
+        self, circuit: QuantumCircuit | DAGCircuit
```

Comment:
This looks to me like a code smell. Can't we have a more generic entity that works as an umbrella for these two types of circuits?

Generic code should lead to simpler code.

Code like this should lead to if/elses, checks in the implementation.

Reply:
You explained this to me already. `QuantumCircuit` and `DAGCircuit` are not classes of our own (they come from external libraries such as Qiskit), so we cannot really have them subclass a generic class of ours.